### PR TITLE
flag-o-matic.eclass: a complete rewrite of get-flag()

### PR DESCRIPTION
**Pull request by rindeal**
- fix case:
  - `CFLAGS='-O1 -O2'`
  - `get-flag '-O*'`
  - before `-O1`
  - now `-O2`
- fix case:
  - `CFLAGS='-W1,-O1'`
  - `get-flag '-O*'`
  - before `-W1,O1`
  - now return 1

`get-flag march` == "i686" syntax still works.

### CODE REVIEW

**Comment by monsieurp** [comment]
@rindeal, hi. you know the drill by now: has this change been reviewed on the gentoo-dev ml?

**Comment by rindeal** [comment]
Hi @monsieurp,

https://devmanual.gentoo.org/eclass-writing/ states that such workflow is necessary for major changes, API changes and new eclasses. None of those conditions apply here.

**Comment by monsieurp** [comment]
sure but they ought to get reviewed nonetheless by someone from the toolchain team. @blueness @vapier @dirtyepic

**Comment by monsieurp** [comment]
also I see bugfix in your commit message. Could you please provide a bug number? thanks!

**Comment by rindeal** [comment]
> also I see bugfix in your commit message. Could you please provide a bug number? thanks!

I'm not aware of any bug in bugzilla covering these bugs.

**Comment by mgorny** [comment]
@rindeal, it is also necessary for widely-used eclasses. Please `git send-email --to gentoo-dev@lists.gentoo.org` ;-).

File: eclass/tests/flag-o-matic.sh [hunk 1]
```diff
@@ -146,4 +146,74 @@ out=$(CC=clang test-flags-CC -finline-limit=1200)
 ftend
 fi
 
+clear-env() {
+	unset $(all-flag-vars)
+}
+
+fom-tbegin() {
+	clear-env
```
Comment (mgorny) [inline eclass/tests/flag-o-matic.sh]: Couldn't you just `clear-env` once and limit the scope of `CFLAGS=` being set for calls?

Comment (mgorny) [inline eclass/tests/flag-o-matic.sh]: (like `CFLAGS='-foo' fom-run ...)

Comment (rindeal) [inline eclass/tests/flag-o-matic.sh]: I've had this idea as well, unfortunately after I've written all of it and haven't considered it to be worth the effort to change it.

**Comment by rindeal** [comment]
I've updated the following:

- more comments, almost every line is now commented
- more space between expressions for an improved reading experience
- needle is computed only once
- comparison now uses a comparison operator instead of a parameter expansion

All in all, the code is now greater than ever.

**Comment by monsieurp** [comment]
@mgorny is this PR ready for prime time?

**Comment by mgorny** [comment]
Was the final version posted for ml review?

**Comment by rindeal** [comment]
[There you go](https://archives.gentoo.org/gentoo-dev/message/1ab348652db236e9c8d670d3dbb491ef)

**Comment by BalkanMadman** [comment]
Let's get this merged. Do you need help sending the patchset to the mailing list?

**Comment by rindeal** [comment]
@BalkanMadman I've stopped using gentoo a long time ago. I don't know why I reopened and rebased this in 2024, probably was bored or something 😁. Feel free to send this code wherever you want... 😉

**Comment by BalkanMadman** [comment]
Quite a coincidence, but this has actually been picked up by Eli and Sam (Gentoo devs) here: https://public-inbox.gentoo.org/gentoo-dev/20260127081259.177126-2-eschwartz@gentoo.org/

Thank you! :)

**Comment by rindeal** [comment]
> we also do odd things like match `get-flag -g` to retrieve `-grecord-gcc-switches` or `-fno-gnu-keywords` etc.

> something that has always bothered me with eclass making incompatible changes on EAPI bump is that afaik they're often not really documented anywhere or the documentation is scattered.  

> It is also a bit awkward to do "-flag=" and then trim off the flag name by hand.  

> then there's also changes where it may silently fail some condition and cause bigger problems that aren't always obvious to spot until a bug report.  

> didn't even know it could match substrings and it sounds ridiculous.  

> Someone before me had used the "omit the -" behaviour which I did not really know about (doesn't help that the flag-o-matic eclass docs do not document it and it's instead in some comment inside the get-flag() function). Then when I changed things around with wrong assumptions it broke things for some flags and I overlooked it while focused on other changes.


😂🤣 oh, Gentoo will never change... thanks for the link, it gave me a good laugh

**Pull request closed without merging**